src/tokenRefresh.ts

```diff
@@ -1,41 +1,41 @@
 import apiClient from "./apiClient"
-import { Service, postData } from "./base"
-
-const refreshToken = async () => {
-  const { refresh_token, access_token } = await postData(
-    Service.TOKEN_REFRESH,
-    "refresh-token",
-    { token: apiClient.tokens.refreshToken }
-  )
-
-  apiClient.setTokens({
-    accessToken: access_token,
-    refreshToken: refresh_token,
-  })
+
+const invokeTokenRefresh = async () => {
+  const { accessToken } = await apiClient.refreshToken()
+
+  apiClient.setAccessToken(accessToken)
   if (apiClient.handlers.onAccessTokenUpdate) {
-    apiClient.handlers.onAccessTokenUpdate(apiClient.tokens.accessToken)
+    apiClient.handlers.onAccessTokenUpdate(apiClient.accessToken)
   }
 }
 
+let pendingRefreshRequest = false
+const waitForPendingRefresh = async (apiCall) => {
+  return new Promise((resolve) => {
+    setTimeout(async () => {
+      return pendingRefreshRequest
+        ? resolve(waitForPendingRefresh(apiCall))
+        : resolve(apiCall())
+    }, 100)
+  })
+}
+
 const refreshTokenAndRetry = async (apiCall) => {
+  if (pendingRefreshRequest) return waitForPendingRefresh(apiCall)
+
   try {
-    await refreshToken()
-    return apiCall()
+    pendingRefreshRequest = true
```

Comment:
```suggestion
    hasPendingRefreshRequest = true
```

to indicate with the name that it's a "state" and not a "request"?

Reply:
really optional. the types might point this out already

Reply:
Fair point, will update it manually to keep the code running